### ISSUE TITLE
Add pipelined batching to the HTTP server using BatchGenerator (throughput and latency improvements)

### ISSUE THREAD

**Issue by billm** [billm/mlx-lm#1]
## Summary
The current HTTP server handles one request per model loop using per-request generation. While `batch_generate` and `BatchGenerator` exist, they aren’t integrated into the server to enable concurrent, pipelined batching of multiple client requests. This issue proposes adding a central batching worker (BatchWorker) to the server that uses `BatchGenerator.insert()` and `next()` to interleave prompt prefill and token generation across active requests, improving throughput without sacrificing streaming latency.

## Motivation
- **Throughput:** Efficiently utilize device by decoding multiple sequences together.
- **Latency:** Admit new requests continuously; do not require a static batch at start.
- **Streaming:** Preserve per-request streaming; tokens should flow as soon as they’re produced.
- **Compatibility:** Keep server API the same by default; add a `--batched` mode via CLI flags.

## Key observation
`BatchGenerator` already supports pipelined insertion:
- `insert(prompts, max_tokens)` queues new requests.
- `next()` interleaves prompt prefill and completion steps, growing/shrinking the active batch as items start/finish.
Therefore we can build a server-side scheduler around `BatchGenerator` without a deep rewrite.

## High-level design
- Add a **BatchWorker** that owns a single `BatchGenerator` and drives it in a dedicated thread:
    - Accept submissions from request handlers (tokenized prompt, per-request max_tokens, and a client context).
    - Call `insert()` as requests arrive.
    - Loop over `next()` to get per-uid tokens; dispatch them back to the request that owns the uid.
    - Stop when a request’s finish condition is met or the client disconnects.

- Keep the HTTP handler thin:
    - Tokenize prompt and compute per-request max_tokens.
    - Submit to BatchWorker; stream tokens to the client as they arrive (SSE or chunked HTTP).
    - Apply per-request stop sequences in the handler; when matched, signal cancellation to the worker.

- **Per-request sampling options:**
    - Initial version: one worker per “sampler config” (temperature, top_p, top_k, etc.). Route requests to the worker whose config matches the server’s flags.
    - Future: per-uid sampler/logits processors inside BatchGenerator.

### CLI additions
- `--batched`: enable batched serving (off by default).
- `--completion-batch-size <int>`: maximum concurrent decoding batch.
- `--prefill-batch-size <int>`: how many new prompts to prefill when capacity is available.
- `--prefill-step-size <int>`: chunk size for long prompt prefill.
- `--batch-max-wait-ms <int>` (optional): upper bound wait time to accumulate a micro-batch when idle.

### 1) Extend BatchGenerator with cancel support (minimal, surgical change)
**Goal:** Allow the server to stop work for specific uids when a request completes early (e.g., stop sequence matched, client disconnected).

**Changes:**
- Add a `cancel(self, uids: list[int])` method:
    - Remove matching entries from `self.unprocessed_prompts`.
    - Filter out matching entries from `self.active_batch` if present.

**Sketch:**
- Track a set `self._canceled: set[int]` to cheaply skip canceled uids that might reappear within an iteration.
- In `_next()`, after computing responses and before filtering `end_idx`/`keep_idx`, check if any uids are in `_canceled` and treat them as finished (drop from batch).
- Ensure `Batch.filter` already updates uids, y, logprobs, max_tokens, num_tokens, cache coherently.

```python
# mlx_lm/generate.py
class BatchGenerator:
    def __init__(...):
        ...
        self._canceled = set()

    def cancel(self, uids: list[int]):
        self._canceled.update(uids)
        self.unprocessed_prompts = [t for t in self.unprocessed_prompts if t[0] not in self._canceled]
        if self.active_batch is not None:
            keep_idx = [i for i, uid in enumerate(self.active_batch.uids) if uid not in self._canceled]
            if keep_idx:
                self.active_batch.filter(keep_idx)
            else:
                self.active_batch = None

    def _next(self):
        ...
        # After computing responses and end_idx/keep_idx:
        for i, uid in enumerate(batch.uids):
            if uid in self._canceled and i not in end_idx:
                end_idx.append(i)
        # Optionally clear from self._canceled
        self._canceled.difference_update([batch.uids[i] for i in end_idx])
        ...
```

### 2) Add BatchWorker to server.py
**Responsibilities:**
- Own BatchGenerator(model, stop_tokens=tokenizer.eos_token_ids, sampler=..., completion_batch_size=..., prefill_batch_size=..., prefill_step_size=...).
- Maintain a mapping from uid -> ClientContext:
    - ClientContext contains queues or async callbacks for token events, finish signals, and per-request stop sequence state.
- Provide:
    - submit(prompt_ids: list[int], max_tokens: int, ctx: ClientContext) -> uid
    - cancel(uid: int)
- Thread:
    - Wrap the loop in wired_limit(model, [generation_stream])
    - while running:
        - responses = gen.next()
        - For each r in responses:
            - ctx = uid_to_ctx[r.uid]
            - If r.finish_reason is None: deliver token/logprobs to ctx else: mark finished; close ctx stream; drop uid from mapping

```python
# mlx_lm/server.py (sketch)
class BatchWorker(Thread):
    def __init__(self, model, tokenizer, sampler, completion_batch_size, prefill_batch_size, prefill_step_size):
        ...
        self.gen = BatchGenerator(...)
        self.uid_to_ctx = {}
        self.lock = threading.Lock()
    def submit(self, prompt_ids, max_tokens, ctx):
        with self.lock:
            uids = self.gen.insert([prompt_ids], max_tokens)
            uid = uids[0]
            self.uid_to_ctx[uid] = ctx
        return uid
    def cancel(self, uid):
        with self.lock:
            self.gen.cancel([uid])
            if uid in self.uid_to_ctx:
                del self.uid_to_ctx[uid]
    def run(self):
        with wired_limit(self.gen.model, [generation_stream]):
            while True:
                responses = self.gen.next()
                for r in responses:
                    ctx = self.uid_to_ctx.get(r.uid)
                    if ctx is None:
                        continue
                    if r.finish_reason is None:
                        ctx.enqueue_token(r.token, r.logprobs)
                    else:
                        ctx.signal_finish(r)
                        self.cancel(r.uid)
```

### 3) Wire HTTP handlers to BatchWorker (only when --batched is set)
- At server startup:
    - Build sampler = make_sampler(...) from CLI flags.
    - Create a singleton BatchWorker with those flags.
- In APIHandler:
    - Tokenize prompt as it does today.
    - Resolve per-request max_tokens and eos tokens from tokenizer.
    - Submit to BatchWorker; get uid and a per-request queue/iterator.
    - If streaming:
        - Flush tokens as they arrive.
        - Track stop sequences in the handler (multi-token sequences may span tokens; keep a rolling suffix buffer).
        - If matched, call worker.cancel(uid).
    - If not streaming:
        - Accumulate tokens until finish and return concatenated text.

## Acceptance criteria
- With `--batched` enabled, concurrent clients receive streamed tokens while the model runs with an active batch > 1 when multiple requests are in flight.
- Requests can arrive at any time; they start prefill promptly without waiting for a static batch.
- Server supports early stopping via both EOS and custom sequences, and cancels work for completed requests.
- Performance improves vs. non-batched mode for multiple concurrent requests on supported hardware.
